NebulaModel/Networking/NebulaConnection.cs

```diff
@@ -24,25 +27,47 @@ public NebulaConnection(WebSocket peerSocket, EndPoint peerEndpoint, NetPacketPr
 
         public void SendPacket<T>(T packet) where T : class, new()
         {
-            if (peerSocket.ReadyState == WebSocketState.Open)
-            {
-                peerSocket.Send(packetProcessor.Write(packet));
-            }
-            else
+            lock (pendingPackets)
             {
-                Log.Warn($"Cannot send packet {packet?.GetType()} to a closed connection {peerEndpoint}");
-            }
+                byte[] rawData = packetProcessor.Write(packet);
+                pendingPackets.Enqueue(rawData);
+                ProcessPacketQueue();
+            }            
         }
 
         public void SendRawPacket(byte[] rawData)
         {
+            lock (pendingPackets)
+            {
+                pendingPackets.Enqueue(rawData);
+                ProcessPacketQueue();
+            }            
+        }
+
+        public void ProcessPacketQueue()
+        {
+            if (pendingPackets.Count == 0 || !enable)
+            {
+                return;
+            }
+            byte[] packet = pendingPackets.Dequeue();
```

Comment:
This needs to be run under the lock.

Either make the method private or take while using the collection.

Alternatively, you can use a ConcurrentQueue.

Reply:
But I've already enclosed all 3 positions calling ``ProcessPacketQueue()`` in the lock.  Should I find another way to write these functions?  
What I hope to achieve is to minimum locks use and also preserve the packet oreder.

Reply:
Yeah, but the API is public is what I was getting to. Anything can end up calling this, and it wouldn't necessarily have the lock taken. You could just make the method private.